_episodes/07-b-numpy.md

```diff
@@ -0,0 +1,280 @@
+data = distances[1:]
+data = data.astype(numpy.float)
+~~~
+{: .language-python}
+
+In a previous lesson, we calculated the mean of each column of the array using the `range` function and a `for` loop. This was good because it reminded us of the `range` function and how `for` loops worked. However, the `numpy.mean` function will let us do that without a for loop.
+
+Checkout the [numpy documentation](https://docs.scipy.org/doc/numpy/reference/generated/numpy.mean.html) for the mean function. Can you figure out what argument we might change to get the mean of each column?
+
+We will use the `axis` argument. If we do not specify `axis`, the mean of the entire array is computed. For the `axis` argument, rows correspond to axis 0, and columns correspond to axis 1. This is similar to slices where you give the indices as `row, column`.
+
+~~~
+column_averages = np.mean(data, axis=0)
+print(column_averages)
+~~~
+{: .language-python}
+
+~~~
+[5000.5          10.87695093    7.34234496   11.20979133   10.9934435 ]
+~~~
+{: .output}
+
+> ## Check your understanding
+> How would you use the `mean` function and slicing to only get the mean of the columns of data (ie, exluding the frame). Save your answer
+> as a variable called `data_averages`.
+>> ## Answer
+>> We would add slicing to `data` in our mean calculation.
+>> ~~~
+>> data_averages = np.mean(data[:,1:], axis=0)
+>> print(data_averages)
+>> ~~~
+>> ~~~
+>> [10.87695093  7.34234496 11.20979133 10.9934435 ]
+>> ~~~
+>> {: .output}
+>> {: .language-python}
+> {: .solution}
+{: .challenge}
+
+## Broadcasting
+
+Another special thing about numpy is something called **broadcasting**. Broadcasting occurs when you attempt mathematical operations on arrays that have different shapes. If possible, the smaller array is "broadcast" across the larger array.
+
+For example, using numpy arrays, you can multiply every element in a numpy array by 10
+
+~~~
+print(10*r1_array)
+~~~
+{: .language-python}
+
+~~~
+[0. 1. 0.]
+~~~
+{: .output}
+
+But, if you multiply our variable `r1` (which is a list), by 10, the list will just be repeated 10 times.
+
+~~~
+print(r1*10)
+~~~
+{: .language-python}
+
+~~~
+[0.0, 0.1, 0.0, 0.0, 0.1, 0.0, 0.0, 0.1, 0.0, 0.0, 0.1, 0.0, 0.0, 0.1, 0.0, 0.0, 0.1, 0.0, 0.0, 0.1, 0.0, 0.0, 0.1, 0.0, 0.0, 0.1, 0.0, 0.0, 0.1, 0.0]
+~~~
+{: .output}
+
+The multiplication of `r1_array*10` is an example of what is called *broadcasting* in NumPy. You can think of the scalar `10` being stretched during the operation to be the same size and shape as `r1_array`.
+
+Broadcasting looks at the dimensions of each array, and sees if it can match them. For example, if you attempt array addition with an array that has size `(1000, 4)`, and the other has `(1, 4)`, `numpy` will notice that each has four columns, and *broadcast* the smaller array over the larger one.
+
+Imagine you have molecular coordinates stored in a numpy array, and you'd like to translate them by a particular distance vector.
+
```

Comment:
We could note here that this is also how arrays work in MATLAB, if we think that is something students would have familiarity with.  I have never heard it called broadcasting in MATLAB, but who knows.  We could also just not say this at all.